src/coreclr/jit/lower.cpp

```diff
@@ -5551,6 +5561,20 @@ GenTree* Lowering::LowerDelegateInvoke(GenTreeCall* call)
     BlockRange().Remove(thisArgNode);
     BlockRange().InsertBefore(call, newThisAddr, newThis, thisArgNode);
 
+#if !defined(TARGET_XARCH)
+    if (comp->GetInterruptible())
+    {
+        // If the target's backend doesn't support indirect calls with immediate operands (contained)
+        // and the method is marked as interruptible, we need to insert a GT_START_NONGC before the call.
+        // to keep the delegate object alive while we're obtaining the function pointer.
+        GenTree* startNonGCNode = new (comp, GT_START_NONGC) GenTree(GT_START_NONGC, TYP_VOID);
+        BlockRange().InsertAfter(thisArgNode, startNonGCNode);
+
+        // We should try to end the non-GC region just before the actual call.
+        *shouldEndNogcBeforeCall = true;
+    }
+#endif
+
```

Comment:
I think it would be cleaner to insert both `START` and `STOP` at the end of `LowerCall` instead of doing half in here and half in `LowerCall`.

Also, is inserting the stop node before the call node really correct?

Reply:
> Also, is inserting the stop node before the call node really correct?

Why not?

Reply:
The GC is illegal when we are on the `blr` instruction -- it is not illegal when we are on the `ldr` instruction. https://www.diffchecker.com/NwHIxdIA/ looks like it makes the GC illegal on the `ldr` instruction but not on the `blr` instruction.

Reply:
can GC happen on call instructions before we entered the call?
If I need to extend the nogc past the call then it's easier for my PR, I just though that I should not add a call to nogc block (because GC may happens inside the call)

Reply:
Addressed

Reply:
> can GC happen on call instructions before we entered the call?

Yes, I think that's the problem here. If GC happens on that call instruction nothing indicates the delegate is live and thus the target we are going to jump to may be collected.

> I just though that I should not add a call to nogc block (because GC may happens inside the call)

Hmm, I'm not totally sure what happens, but the safepoint should logically be on the return address, not on the call, so I think we should be fine.

Reply:
I was just wondering if we have any debug mechanism that checks that all calls inside nogc blocks can't trigger GC even internally, but I presume we don't otherwise we'd catch the recent bug with BULKBARRIER earlier